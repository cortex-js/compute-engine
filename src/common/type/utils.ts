import { parseType } from './parse';
import { PRIMITIVE_TYPES } from './primitive';
import { typeToString } from './serialize';
import { isSubtype } from './subtype';
import type {
  Type,
  FunctionSignature,
  TypeString,
  PrimitiveType,
} from './types';

/** Given two types a and b, return the narrowest type common to a and b */
function narrow2(a: Readonly<Type>, b: Readonly<Type>): Readonly<Type> {
  if (a === b) return a;

  if (a === 'nothing' || b === 'nothing') return 'nothing';

  if (a === 'any') return b;
  if (b === 'any') return a;

  if (a === 'never') return b;
  if (b === 'never') return a;

  if (a === 'unknown') return b;
  if (b === 'unknown') return a;

  if (isSubtype(a, b)) return a;
  if (isSubtype(b, a)) return b;

  return superType(a, b);
}

/** Given two types, return the broadest  */
function widen2(a: Readonly<Type>, b: Readonly<Type>): Readonly<Type> {
  if (a === b) return a;
  if (a === 'any' || b === 'any') return 'any';

  if (a === 'never') return b;
  if (b === 'never') return a;

  if (a === 'unknown') return b;
  if (b === 'unknown') return a;

  if (a === 'nothing') return b;
  if (b === 'nothing') return a;

  if (isSubtype(a, b)) return b;
  if (isSubtype(b, a)) return a;

  return superType(a, b);
}

/** Convert two or more types into a more specific type that is a subtype of
 *  all the input types. The resulting type is usually more constrained and
 *  only encompasses values that belong to both input types.
 *
 * Examples:
 * narrow('integer', 'rational') => 'integer'
 * narrow('number', 'complex') => 'complex'
 * narrow('number', 'collection') => 'nothing'
 * narrow('number', 'value') => 'value'
 * narrow('number', 'expression') => 'expression'
 * narrow('number', 'string') => 'nothing'
 *
 *
 */
export function narrow(...types: Readonly<Type>[]): Type {
  if (types.length === 0) return 'nothing';
  if (types.length === 1) return types[0];

  return types.reduce((a, b) => narrow2(a, b));
}

/**
 * Convert two or more types into a broader, more general type that can
 * accommodate all the input types. The resulting type is usually a supertype
 * that encompasses the possible values of the input types
 *
 * Examples:
 * widen('integer', 'rational') => 'rational'
 * widen('number', 'complex') => 'complex'
 * widen('number', 'collection') => 'collection'
 * widen('number', 'value') => 'value'
 * widen('number', 'expression') => 'expression'
 * widen('number', 'string') => 'any'
 */
export function widen(...types: Readonly<Type>[]): Readonly<Type> {
  if (types.length === 0) return 'nothing';
  if (types.length === 1) return types[0];

  return types.reduce((a, b) => widen2(a, b));
}

export function isSignatureType(
  type: Readonly<Type> | TypeString
): type is FunctionSignature {
  type = typeof type === 'string' ? parseType(type) : type;
  return typeof type !== 'string' && type.kind === 'signature';
}

export function functionSignature(type: Readonly<Type>): Type | undefined {
  if (type === 'function') return parseType('(any*) -> unknown');
  if (typeof type === 'string') return undefined;

  if (type.kind === 'signature') return type;
  return undefined;
}

export function functionResult(
  type: Readonly<Type> | undefined
): Type | undefined {
  if (!type) return undefined;
  if (type === 'function') return 'any';
  if (typeof type === 'string') return undefined;
  if (type.kind === 'signature') return type.result;
  return undefined;
}

export function collectionElementType(type: Readonly<Type>): Type | undefined {
  if (type === 'collection') return 'any';
  if (type === 'indexed_collection') return 'any';
  if (type === 'list') return 'any';
  if (type === 'set') return 'any';
  if (type === 'tuple') return 'any';
  if (type === 'dictionary') return 'any';
  if (type === 'record') return 'any';
  if (typeof type === 'string') return undefined;

  if (type.kind === 'collection' || type.kind === 'indexed_collection')
    return type.elements;

  if (type.kind === 'list') return type.elements;

  if (type.kind === 'set') return type.elements;

  if (type.kind === 'tuple') return widen(...type.elements.map((x) => x.type));

  if (type.kind === 'dictionary')
    return parseType(`tuple<string, ${type.values}>`);

  if (type.kind === 'record') {
    return parseType(
      `tuple<string, ${typeToString(widen(...Object.values(type.elements)))}>`
    );
  }

  return undefined;
}

export function isValidTypeName(name: string): boolean {
  return /^[a-zA-Z_][a-zA-Z0-9_]*$/.test(name);
}

export function isValidType(t: any): t is Readonly<Type> {
  if (typeof t === 'string')
    return PRIMITIVE_TYPES.includes(t as PrimitiveType);
  if (typeof t !== 'object') return false;
  if (!('kind' in t)) return false;
  return (
    t.kind === 'signature' ||
    t.kind === 'union' ||
    t.kind === 'intersection' ||
    t.kind === 'negation' ||
    t.kind === 'tuple' ||
    t.kind === 'list' ||
    t.kind === 'record' ||
    t.kind === 'dictionary' ||
    t.kind === 'set' ||
    t.kind === 'function' ||
    t.kind === 'collection' ||
    t.kind === 'indexed_collection' ||
    t.kind === 'reference'
  );
}

function superType(a: Readonly<Type>, b: Readonly<Type>): Type {
  // Return the common super type of a and b
  if (a === b) return a;
  if (a === 'any' || b === 'any') return 'any';
  if (a === 'never') return b;
  if (b === 'never') return a;
  if (a === 'unknown') return b;
  if (b === 'unknown') return a;
  if (a === 'nothing') return b;
  if (b === 'nothing') return a;

  // Check in order from most specific to most general
  if (commonSupertype(a, b, 'non_finite_number')) return 'non_finite_number';

  if (commonSupertype(a, b, 'finite_integer')) return 'finite_integer';
  if (commonSupertype(a, b, 'integer')) return 'integer';
  if (commonSupertype(a, b, 'finite_rational')) return 'finite_rational';
  if (commonSupertype(a, b, 'rational')) return 'rational';
  if (commonSupertype(a, b, 'finite_real')) return 'finite_real';
  if (commonSupertype(a, b, 'real')) return 'real';

  if (commonSupertype(a, b, 'imaginary')) return 'imaginary';

  if (commonSupertype(a, b, 'finite_complex')) return 'finite_complex';
  if (commonSupertype(a, b, 'complex')) return 'complex';

  if (commonSupertype(a, b, 'finite_number')) return 'finite_number';
  if (commonSupertype(a, b, 'number')) return 'number';

  if (commonSupertype(a, b, 'list')) return 'list';
  if (commonSupertype(a, b, 'record')) return 'record';
  if (commonSupertype(a, b, 'dictionary')) return 'dictionary';
  if (commonSupertype(a, b, 'set')) return 'set';
  if (commonSupertype(a, b, 'tuple')) return 'tuple';
  if (commonSupertype(a, b, 'indexed_collection')) return 'indexed_collection';
  if (commonSupertype(a, b, 'collection')) return 'collection';

  if (commonSupertype(a, b, 'scalar')) return 'scalar';
  if (commonSupertype(a, b, 'value')) return 'value';
  if (commonSupertype(a, b, 'function')) return 'function';

  if (commonSupertype(a, b, 'expression')) return 'expression';

  return 'any';
}

function commonSupertype(
  a: Readonly<Type>,
  b: Readonly<Type>,
  ancestor: Readonly<Type>
): boolean {
  if (isSubtype(a, ancestor) && isSubtype(b, ancestor)) return true;
  return false;
}
