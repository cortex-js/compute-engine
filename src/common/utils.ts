export function permutations<T>(
  xs: ReadonlyArray<T>
): ReadonlyArray<ReadonlyArray<T>> {
  const result: ReadonlyArray<T>[] = [];

  const permute = (arr, m = []) => {
    if (arr.length === 0) {
      result.push(m);
    } else {
      for (let i = 0; i < arr.length; i++) {
        const curr = arr.slice();
        const next = curr.splice(i, 1);
        permute(curr.slice(), m.concat(next));
      }
    }
  };

  permute(xs);

  return result;
}

export function hidePrivateProperties(obj: any) {
  for (const key in obj) {
    if (key.startsWith('_') && obj.hasOwnProperty(key)) {
      Object.defineProperty(obj, key, {
        enumerable: false,
        configurable: true, // Allows redefinition if necessary
        writable: true, // Allows modification
        value: obj[key],
      });
    }
  }
}
