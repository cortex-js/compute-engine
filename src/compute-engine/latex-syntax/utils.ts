import { joinLatex } from './tokenizer';
import { DEFINITIONS_INEQUALITIES } from './dictionary/definitions-relational-operators';

export function isLatexString(s: unknown): s is string {
  if (typeof s === 'string') return s.startsWith('$') && s.endsWith('$');
  return false;
}

export function asLatexString(s: unknown): string | null {
  if (typeof s === 'number') return s.toString();
  if (typeof s === 'string') {
    const str = s.trim();

    if (str.startsWith('$$') && str.endsWith('$$')) return str.slice(2, -2);
    if (str.startsWith('$') && str.endsWith('$')) return str.slice(1, -1);
  }
  if (Array.isArray(s)) {
    // Check after 'string', since a string is also an array...
    return asLatexString(joinLatex(s));
  }
  return null;
}

export function isRelationalOperator(name: string | undefined): boolean {
  if (typeof name !== 'string') return false;
  return DEFINITIONS_INEQUALITIES.some((x) => x.name === name);
}

export function isInequalityOperator(operator: string | undefined): boolean {
  if (typeof operator !== 'string') return false;
  return ['Less', 'LessEqual', 'Greater', 'GreaterEqual'].includes(operator);
}

export function isEquationOperator(operator: string | undefined): boolean {
  if (typeof operator !== 'string') return false;
  return ['Equal', 'NotEqual'].includes(operator);
}
