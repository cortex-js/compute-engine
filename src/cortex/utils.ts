import { ErrorSignal, Signal } from '../common/signals.ts';

export class CortexError {
  signal: ErrorSignal;
  constructor(errorSignal: Signal) {
    this.signal = { severity: 'error', ...errorSignal } as ErrorSignal;
  }
  toString(): string {
    let result = '';
    if (this.signal.head) result += this.signal.head + ': ';

    if (typeof this.signal.message === 'string') {
      result += this.signal.message;
    } else {
      result += ' ';
      for (const arg of this.signal.message) result += arg.toString() + ' ';
    }

    return result;
  }
}
