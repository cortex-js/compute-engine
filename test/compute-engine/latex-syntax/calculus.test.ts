import { engine as ce, exprToString } from '../../utils';

function parse(s: string) {
  const expr = ce.parse(s);
  return expr.toLatex() + '\n' + expr.toString() + '\n' + exprToString(expr);
}

/** Integrals with no limits specified */
describe('INDEFINITE INTEGRAL', () => {
  test('No index', () => {
    expect(parse('\\int\\sin x + 1 = 2')).toMatchSnapshot();
  });

  test('Index with d', () => {
    expect(parse('\\int\\sin x \\operatorname{d} x+1 = 2')).toMatchSnapshot();
  });

  test('Index with mathrm', () => {
    expect(parse('\\int\\sin x \\mathrm{d}x+1 = 2')).toMatchSnapshot();
  });

  test('Index with operatorname', () => {
    expect(parse('\\int\\sin x \\operatorname{d}x+1 = 2')).toMatchSnapshot();
  });

  test('Index \\alpha', () => {
    expect(parse('\\int\\alpha d\\alpha+1 = 2')).toMatchSnapshot();
  });

  test('Spacing commands', () => {
    expect(parse('\\int\\!\\sin x \\,\\mathrm{d}x+1 = 2')).toMatchSnapshot();
  });

  test('Index in addition', () =>
    expect(parse('\\int3x+kxdx = 2')).toMatchSnapshot());

  test('Index in negate', () =>
    expect(parse('\\int-xdx = 2')).toMatchSnapshot());

  test('Index inside delimiter', () =>
    expect(parse('\\int(3x+x^2dx) = 2')).toMatchSnapshot());

  test('Index after delimiter', () =>
    expect(parse('\\int(3x+x^2)dx = 2')).toMatchSnapshot());

  test('Index after implicit trig argument', () =>
    expect(parse('\\int\\sin x dx = 2')).toMatchSnapshot());

  test('Index in numerator', () =>
    expect(parse('\\int\\frac{3xdx}{5} = 2')).toMatchSnapshot());

  test('Index in numerator with mathrm', () =>
    expect(parse('\\int\\frac{3x\\mathrm{d}x}{5} = 2')).toMatchSnapshot());

  test('Index with \\differentialD', () =>
    expect(parse('\\int\\sin x \\differentialD x + 1 = 2')).toMatchSnapshot());

  test('Non-standard typesetting (\\cdot)', () =>
    expect(parse('\\int f(t) \\cdot dt')).toMatchSnapshot());

  test('INVALID index in denominator', () =>
    expect(parse('\\int\\frac{3x}{5dx} = 2')).toMatchSnapshot()); // @fixme, should error
});

/** Integral with limits */
describe('DEFINITE INTEGRAL', () => {
  test('Lower and upper bounds', () => {
    expect(parse('\\int_0^1\\sin x dx')).toMatchSnapshot();
  });

  test('Lower and upper bounds with \\limits', () => {
    expect(parse('\\int\\limits_0^1\\sin x dx')).toMatchSnapshot();
  });

  test('Lower bound only (malformed)', () => {
    expect(parse('\\int_0\\sin x dx')).toMatchSnapshot();
  });

  test('Upper bound only (malformed)', () => {
    expect(parse('\\int^1\\sin x dx')).toMatchSnapshot();
  });

  test('Symbolic lower bound (regions)', () => {
    expect(parse('\\int_G\\sin x dx')).toMatchSnapshot();
  });

  test('Lower and upper bounds with no index', () => {
    expect(parse('\\int_0^1\\sin x')).toMatchSnapshot();
  });
});

describe('MULTIPLE INTEGRALS', () => {
  test('Double integral', () => {
    expect(parse('\\int_1^2\\int_0^1 x^2+y^2 dx dy')).toMatchSnapshot();
  });

  test('Double integral with repeated index', () => {
    expect(parse('\\int_1^2\\int_0^1 x^2 dx dx')).toMatchSnapshot();
  });

  test('Triple integral', () => {
    expect(
      parse('\\int_1^2\\int_0^1\\int_3^4 x^2+y^2+z^2 dx dy dz')
    ).toMatchSnapshot();
  });

  test('Triple integral with \\limits and spacing', () => {
    expect(
      parse(
        '\\int\\limits_1^2\\int\\limits_0^1\\int\\limits_3^4 \\!x^2+y^2+z^2 \\,\\mathrm{d}x \\mathrm{d}y \\mathrm{d}z'
      )
    ).toMatchSnapshot();
  });
});

describe('EXOTIC INTEGRALS', () => {
  test('\\oint - contour integral', () => {
    expect(parse('\\oint_V f(s) ds')).toMatchSnapshot();
  });

  test('\\oint - contour integral, no index', () => {
    expect(parse('\\oint_C f')).toMatchSnapshot();
  });

  test('\\intop', () => {
    expect(parse('\\intop_0^1\\sin x dx')).toMatchSnapshot();
  });

  test('\\smallint', () => {
    expect(parse('\\smallint_0^1\\sin x dx')).toMatchSnapshot();
  });

  test('\\iint', () => {
    expect(parse('\\iint_D f(x,y) dx dy')).toMatchSnapshot();
  });
  test('\\iiint', () => {
    expect(parse('\\iiint_V f(x,y,z) dx dy dz')).toMatchSnapshot();
  });
});

describe('REAL WORLD INTEGRALS', () => {
  test('Integral with non standard typesetting', () =>
    expect(
      parse('S_t=S_0+\\int_{t_i}^{t_e}\\left(G-F\\right)\\cdot dt')
    ).toMatchSnapshot());
});

describe('LIMIT', () => {
  expect(parse('\\lim_{x\\to\\infty} f(x)')).toMatchSnapshot(
    `["Limit", ["Function", ["f", "x"], "x"], "PositiveInfinity"]`
  );

  expect(ce.parse('\\lim_{x \\to 0} \\frac{\\sin(x)}{x}')).toMatchSnapshot(
    `["Limit", ["Function", ["Divide", ["Sin", "x"], "x"], "x"], 0]`
  );
  expect(parse('\\lim_{x\\to\\infty} f(x)')).toMatchSnapshot(`
  lim_(+oo) {f(x)}
  \\lim_{x\\to\\infty}f(x)
  Limit,Function,f,x,x,PositiveInfinity
`);
});
