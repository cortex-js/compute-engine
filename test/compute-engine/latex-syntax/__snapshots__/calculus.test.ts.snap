// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
\\lim_{x\\to\\infty}f(x)
lim_(+oo) {f(x)}
["Limit", ["Function", ["f", "x"], "x"], "PositiveInfinity"]
`;

exports[` 2`] = `["Limit", ["Function", ["Divide", ["Sin", "x"], "x"], "x"], 0]`;

exports[` 3`] = `
\\lim_{x\\to\\infty}f(x)
lim_(+oo) {f(x)}
["Limit", ["Function", ["f", "x"], "x"], "PositiveInfinity"]
`;

exports[`DEFINITE INTEGRAL Lower and upper bounds 1`] = `
\\int_{0}^{1}\\!\\sin(x)\\, \\mathrm{d}x
int_(0)^(1)(sin(x) dx)
["Integrate", ["Function", ["Sin", "x"], "x"], ["Limits", "x", 0, 1]]
`;

exports[`DEFINITE INTEGRAL Lower and upper bounds with \\limits 1`] = `
\\int_{0}^{1}\\!\\sin(x)\\, \\mathrm{d}x
int_(0)^(1)(sin(x) dx)
["Integrate", ["Function", ["Sin", "x"], "x"], ["Limits", "x", 0, 1]]
`;

exports[`DEFINITE INTEGRAL Lower and upper bounds with no index 1`] = `
\\int_{0}^{1}\\!\\sin(x)
int_(0)^(1)(sin(x) dx)
[
  "Integrate",
  ["Function", ["Sin", "x"], "x"],
  ["Limits", "Nothing", 0, 1]
]
`;

exports[`DEFINITE INTEGRAL Lower bound only (malformed) 1`] = `
\\int_{0}\\!\\sin(x)\\, \\mathrm{d}x
int_(0)(sin(x) dx)
[
  "Integrate",
  ["Function", ["Sin", "x"], "x"],
  ["Limits", "x", 0, "Nothing"]
]
`;

exports[`DEFINITE INTEGRAL Symbolic lower bound (regions) 1`] = `
\\int_{G}\\!\\sin(x)\\, \\mathrm{d}x
int_("CatalanConstant")(sin(x) dx)
[
  "Integrate",
  ["Function", ["Sin", "x"], "x"],
  ["Limits", "x", "CatalanConstant", "Nothing"]
]
`;

exports[`DEFINITE INTEGRAL Upper bound only (malformed) 1`] = `
\\int^{1}\\!\\sin(x)\\, \\mathrm{d}x
int^(1)(sin(x) dx)
[
  "Integrate",
  ["Function", ["Sin", "x"], "x"],
  ["Limits", "x", "Nothing", 1]
]
`;

exports[`EXOTIC INTEGRALS \\iiint 1`] = `
\\int\\int\\int_{\\mathtip{\\error{\\blacksquare}}{\\in \\text{unknown}\\notin \\text{number}}}\\!f(x, y, z)\\, \\mathrm{d}x \\mathrm{d}y \\mathrm{d}z
int_(Error(ErrorCode("incompatible-type", "number", "unknown")))(f(x, y, z) dx  dy  dz)
[
  "Integrate",
  ["Function", ["f", "x", "y", "z"], "x", "y", "z"],
  [
    "Limits",
    "x",
    [
      "Error",
      ["ErrorCode", "incompatible-type", "'number'", "'unknown'"]
    ],
    "Nothing"
  ],
  ["Limits", "y", "Nothing", "Nothing"],
  ["Limits", "z", "Nothing", "Nothing"]
]
`;

exports[`EXOTIC INTEGRALS \\iint 1`] = `
\\int\\int_{\\mathtip{\\error{\\blacksquare}}{\\in \\text{(expression, variable: symbol, variables: symbol+) -> expression}\\notin \\text{number}}}\\!f(x, y)\\, \\mathrm{d}x \\mathrm{d}y
int_(Error(ErrorCode("incompatible-type", "number", "(expression, variable: symbol, variables: symbol+) -> expression")))(f(x, y) dx  dy)
[
  "Integrate",
  ["Function", ["f", "x", "y"], "x", "y"],
  [
    "Limits",
    "x",
    [
      "Error",
      [
        "ErrorCode",
        "incompatible-type",
        "'number'",
        "(expression, variable: symbol, variables: symbol+) -> expression"
      ]
    ],
    "Nothing"
  ],
  ["Limits", "y", "Nothing", "Nothing"]
]
`;

exports[`EXOTIC INTEGRALS \\intop 1`] = `
(\\mathrm{Baseform}(\\text{unexpected-command}, \\error{\\blacksquare}),0,\\sin(dxx))
(Baseform("unexpected-command", Error(ErrorCode("invalid-base", "\\"\\intop\\""))), 0, sin(d * x * x))
[
  "Tuple",
  [
    "Baseform",
    "unexpected-command",
    ["Error", ["ErrorCode", "invalid-base", ""\\intop""]]
  ],
  0,
  ["Sin", ["Multiply", "d", "x", "x"]]
]
`;

exports[`EXOTIC INTEGRALS \\oint - contour integral 1`] = `
\\ointundefined\\!f(s)
CircularIntegrate(f(s), (s, V, "Nothing"))
["CircularIntegrate", ["f", "s"], ["Triple", "s", "V", "Nothing"]]
`;

exports[`EXOTIC INTEGRALS \\oint - contour integral, no index 1`] = `
\\ointundefined\\!f
CircularIntegrate(f, ("Nothing", C, "Nothing"))
["CircularIntegrate", "f", ["Triple", "Nothing", "C", "Nothing"]]
`;

exports[`EXOTIC INTEGRALS \\smallint 1`] = `
(\\mathrm{Baseform}(\\text{unexpected-command}, \\error{\\blacksquare}),0,\\sin(dxx))
(Baseform("unexpected-command", Error(ErrorCode("invalid-base", "\\"\\smallint\\""))), 0, sin(d * x * x))
[
  "Tuple",
  [
    "Baseform",
    "unexpected-command",
    ["Error", ["ErrorCode", "invalid-base", ""\\smallint""]]
  ],
  0,
  ["Sin", ["Multiply", "d", "x", "x"]]
]
`;

exports[`INDEFINITE INTEGRAL INVALID index in denominator 1`] = `
\\int\\!\\frac{3x}{5dx}=2
int((3x) / (5d * x) dd  dx) === 2
[
  "Equal",
  [
    "Integrate",
    [
      "Function",
      ["Divide", ["Multiply", 3, "x"], ["Multiply", 5, "d", "x"]],
      "d",
      "x"
    ],
    ["Limits", "Nothing", "Nothing", "Nothing"]
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index \\alpha 1`] = `
\\int\\!\\alpha\\, \\mathrm{d}\\alpha+1=2
int(alpha) + 1 === 2
[
  "Equal",
  [
    "Add",
    ["Integrate", "alpha", ["Limits", "alpha", "Nothing", "Nothing"]],
    1
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index after delimiter 1`] = `
\\int\\!x^2+3x\\, \\mathrm{d}x=2
int(x^2 + 3x dx) === 2
[
  "Equal",
  [
    "Integrate",
    ["Function", ["Add", ["Square", "x"], ["Multiply", 3, "x"]], "x"],
    ["Limits", "x", "Nothing", "Nothing"]
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index after implicit trig argument 1`] = `
\\int\\!\\sin(x)\\, \\mathrm{d}x=2
int(sin(x) dx) === 2
[
  "Equal",
  [
    "Integrate",
    ["Function", ["Sin", "x"], "x"],
    ["Limits", "x", "Nothing", "Nothing"]
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index in addition 1`] = `
\\int\\!kx+3x\\, \\mathrm{d}x=2
int(k * x + 3x dk  dx) === 2
[
  "Equal",
  [
    "Integrate",
    [
      "Function",
      ["Add", ["Multiply", "k", "x"], ["Multiply", 3, "x"]],
      "k",
      "x"
    ],
    ["Limits", "x", "Nothing", "Nothing"]
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index in negate 1`] = `
\\int\\!-x\\, \\mathrm{d}x=2
int(-x dx) === 2
[
  "Equal",
  [
    "Integrate",
    ["Function", ["Negate", "x"], "x"],
    ["Limits", "x", "Nothing", "Nothing"]
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index in numerator 1`] = `
\\int\\!\\frac{3x}{5}\\, \\mathrm{d}x=2
int(3/5 * x dx) === 2
[
  "Equal",
  [
    "Integrate",
    ["Function", ["Multiply", ["Rational", 3, 5], "x"], "x"],
    ["Limits", "x", "Nothing", "Nothing"]
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index in numerator with mathrm 1`] = `
\\int\\!\\frac{3x}{5}\\, \\mathrm{d}x=2
int(3/5 * x dx) === 2
[
  "Equal",
  [
    "Integrate",
    ["Function", ["Multiply", ["Rational", 3, 5], "x"], "x"],
    ["Limits", "x", "Nothing", "Nothing"]
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index inside delimiter 1`] = `
\\int\\!x^2+3x\\, \\mathrm{d}x=2
int(x^2 + 3x dx) === 2
[
  "Equal",
  [
    "Integrate",
    ["Function", ["Add", ["Square", "x"], ["Multiply", 3, "x"]], "x"],
    ["Limits", "x", "Nothing", "Nothing"]
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index with \\differentialD 1`] = `
\\int\\!\\sin(x)\\, \\mathrm{d}x+1=2
int(sin(x) dx) + 1 === 2
[
  "Equal",
  [
    "Add",
    [
      "Integrate",
      ["Function", ["Sin", "x"], "x"],
      ["Limits", "x", "Nothing", "Nothing"]
    ],
    1
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index with d 1`] = `
\\int\\!\\sin(x)\\, \\mathrm{d}x+1=2
int(sin(x) dx) + 1 === 2
[
  "Equal",
  [
    "Add",
    [
      "Integrate",
      ["Function", ["Sin", "x"], "x"],
      ["Limits", "x", "Nothing", "Nothing"]
    ],
    1
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index with mathrm 1`] = `
\\int\\!\\sin(x)\\, \\mathrm{d}x+1=2
int(sin(x) dx) + 1 === 2
[
  "Equal",
  [
    "Add",
    [
      "Integrate",
      ["Function", ["Sin", "x"], "x"],
      ["Limits", "x", "Nothing", "Nothing"]
    ],
    1
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Index with operatorname 1`] = `
\\int\\!\\sin(x)\\, \\mathrm{d}x+1=2
int(sin(x) dx) + 1 === 2
[
  "Equal",
  [
    "Add",
    [
      "Integrate",
      ["Function", ["Sin", "x"], "x"],
      ["Limits", "x", "Nothing", "Nothing"]
    ],
    1
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL No index 1`] = `
\\int\\!0+\\sin(x)+1=2
int(0) + sin(x) + 1 === 2
[
  "Equal",
  [
    "Add",
    [
      "Integrate",
      ["Function", 0],
      ["Limits", "Nothing", "Nothing", "Nothing"]
    ],
    ["Sin", "x"],
    1
  ],
  2
]
`;

exports[`INDEFINITE INTEGRAL Non-standard typesetting (\\cdot) 1`] = `
\\int\\!f(t)\\, \\mathrm{d}t
int(f(t) dt)
[
  "Integrate",
  ["Function", ["f", "t"], "t"],
  ["Limits", "t", "Nothing", "Nothing"]
]
`;

exports[`INDEFINITE INTEGRAL Spacing commands 1`] = `
\\int\\!\\sin(x)\\, \\mathrm{d}x+1=2
int(sin(x) dx) + 1 === 2
[
  "Equal",
  [
    "Add",
    [
      "Integrate",
      ["Function", ["Sin", "x"], "x"],
      ["Limits", "x", "Nothing", "Nothing"]
    ],
    1
  ],
  2
]
`;

exports[`MULTIPLE INTEGRALS Double integral 1`] = `
\\int_{1}^{2}\\!\\int\\int_{0}^{1}\\!x^2+y^2\\, \\mathrm{d}x \\mathrm{d}y
int_(1)^(2)(int_(0)^(1)(x^2 + y^2 dx  dy) dx  dy)
[
  "Integrate",
  [
    "Function",
    [
      "Integrate",
      ["Function", ["Add", ["Square", "x"], ["Square", "y"]], "x", "y"],
      ["Limits", "x", 0, 1],
      ["Limits", "y", "Nothing", "Nothing"]
    ],
    "x",
    "y"
  ],
  ["Limits", "Nothing", 1, 2]
]
`;

exports[`MULTIPLE INTEGRALS Double integral with repeated index 1`] = `
\\int_{1}^{2}\\!\\int\\int_{0}^{1}\\!x^2\\, \\mathrm{d}x \\mathrm{d}x
int_(1)^(2)(int_(0)^(1)(x^2 dx) dx)
[
  "Integrate",
  [
    "Function",
    [
      "Integrate",
      ["Function", ["Square", "x"], "x"],
      ["Limits", "x", 0, 1],
      ["Limits", "x", "Nothing", "Nothing"]
    ],
    "x"
  ],
  ["Limits", "Nothing", 1, 2]
]
`;

exports[`MULTIPLE INTEGRALS Triple integral 1`] = `
\\int_{1}^{2}\\!\\int_{0}^{1}\\!\\int\\int\\int_{3}^{4}\\!x^2+y^2+z^2\\, \\mathrm{d}x \\mathrm{d}y \\mathrm{d}z
int_(1)^(2)(int_(0)^(1)(int_(3)^(4)(x^2 + y^2 + z^2 dx  dy  dz) dx  dy  dz) dx  dy  dz)
[
  "Integrate",
  [
    "Function",
    [
      "Integrate",
      [
        "Function",
        [
          "Integrate",
          [
            "Function",
            ["Add", ["Square", "x"], ["Square", "y"], ["Square", "z"]],
            "x",
            "y",
            "z"
          ],
          ["Limits", "x", 3, 4],
          ["Limits", "y", "Nothing", "Nothing"],
          ["Limits", "z", "Nothing", "Nothing"]
        ],
        "x",
        "y",
        "z"
      ],
      ["Limits", "Nothing", 0, 1]
    ],
    "x",
    "y",
    "z"
  ],
  ["Limits", "Nothing", 1, 2]
]
`;

exports[`MULTIPLE INTEGRALS Triple integral with \\limits and spacing 1`] = `
\\int_{1}^{2}\\!\\int_{0}^{1}\\!\\int\\int\\int_{3}^{4}\\!x^2+y^2+z^2\\, \\mathrm{d}x \\mathrm{d}y \\mathrm{d}z
int_(1)^(2)(int_(0)^(1)(int_(3)^(4)(x^2 + y^2 + z^2 dx  dy  dz) dx  dy  dz) dx  dy  dz)
[
  "Integrate",
  [
    "Function",
    [
      "Integrate",
      [
        "Function",
        [
          "Integrate",
          [
            "Function",
            ["Add", ["Square", "x"], ["Square", "y"], ["Square", "z"]],
            "x",
            "y",
            "z"
          ],
          ["Limits", "x", 3, 4],
          ["Limits", "y", "Nothing", "Nothing"],
          ["Limits", "z", "Nothing", "Nothing"]
        ],
        "x",
        "y",
        "z"
      ],
      ["Limits", "Nothing", 0, 1]
    ],
    "x",
    "y",
    "z"
  ],
  ["Limits", "Nothing", 1, 2]
]
`;

exports[`REAL WORLD INTEGRALS Integral with non standard typesetting 1`] = `
\\mathrm{S_{t}}=\\mathrm{S_0}+\\int_{\\mathtip{\\error{\\blacksquare}}{\\in \\text{unknown}\\notin \\text{number}}}^{\\mathtip{\\error{\\blacksquare}}{\\in \\text{unknown}\\notin \\text{number}}}\\!G-F\\, \\mathrm{d}t
"S_t" === "S_0" + int_(Error(ErrorCode("incompatible-type", "number", "unknown")))^(Error(ErrorCode("incompatible-type", "number", "unknown")))(-F + "CatalanConstant" dF)
[
  "Equal",
  "S_t",
  [
    "Add",
    "S_0",
    [
      "Integrate",
      ["Function", ["Subtract", "CatalanConstant", "F"], "F"],
      [
        "Limits",
        "t",
        [
          "Error",
          ["ErrorCode", "incompatible-type", "'number'", "'unknown'"]
        ],
        [
          "Error",
          ["ErrorCode", "incompatible-type", "'number'", "'unknown'"]
        ]
      ]
    ]
  ]
]
`;
